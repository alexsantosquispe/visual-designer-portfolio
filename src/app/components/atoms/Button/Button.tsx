'use client';

import { twMerge } from 'tailwind-merge';
import cn from 'clsx';

interface ButtonProps {
  title: string;
  type?: 'button' | 'submit';
  onClick?: () => void;
  disabled?: boolean;
}

export const Button = ({
  title,
  type = 'button',
  disabled = false,
  onClick
}: ButtonProps) => {
  return (
    <button
      type={type}
      className={twMerge(
        cn({
          'bg-primary cursor-not-allowed opacity-75': disabled,
          'bg-primary cursor-pointer hover:bg-black': !disabled
        }),
        'w-fit self-center px-8 py-4 text-[1.25rem] font-semibold text-white md:self-auto md:px-16 md:py-6'
      )}
      onClick={typeof onClick === 'function' ? onClick : undefined}
      disabled={disabled}
    >
      {title}
    </button>
  );
};
