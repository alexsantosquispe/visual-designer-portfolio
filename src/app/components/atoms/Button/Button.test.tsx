import { fireEvent, render, screen } from '@testing-library/react';
import { Button } from './Button';

describe('Button component', () => {
  const onClickMock = jest.fn();
  const props = {
    title: 'button title',
    onClick: onClickMock
  };

  describe('Styles', () => {
    it('should render default button correctly', () => {
      let component = render(<Button {...props} />);
      expect(component).toMatchSnapshot();

      component = render(
        <Button {...props} type='submit' onClick={undefined} />
      );
      expect(component).toMatchSnapshot();
    });
  });

  describe('Behavior', () => {
    it('should button be disabled', () => {
      render(<Button {...props} disabled={true} />);

      const button = screen.getByText(props.title);

      expect(button).toBeInTheDocument();

      expect(button).toBeDisabled();
    });

    it('should call onclick', () => {
      render(<Button {...props} />);

      const button = screen.getByText(props.title);

      expect(button).toBeInTheDocument();

      fireEvent.click(button);

      expect(onClickMock).toHaveBeenCalledTimes(1);
    });
  });
});
