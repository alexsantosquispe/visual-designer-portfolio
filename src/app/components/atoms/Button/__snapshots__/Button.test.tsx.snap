// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Styles should render default button correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-label="button title"
        class="bg-primary/90 hover:bg-primary/70 cursor-pointer w-fit self-center rounded-xl px-8 py-4 text-[1.25rem] font-semibold text-white md:self-auto"
        type="button"
      >
        button title
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-label="button title"
      class="bg-primary/90 hover:bg-primary/70 cursor-pointer w-fit self-center rounded-xl px-8 py-4 text-[1.25rem] font-semibold text-white md:self-auto"
      type="button"
    >
      button title
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button component Styles should render default button correctly 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-label="button title"
        class="bg-primary/90 hover:bg-primary/70 cursor-pointer w-fit self-center rounded-xl px-8 py-4 text-[1.25rem] font-semibold text-white md:self-auto"
        type="button"
      >
        button title
      </button>
    </div>
    <div>
      <button
        aria-label="button title"
        class="bg-primary/90 hover:bg-primary/70 cursor-pointer w-fit self-center rounded-xl px-8 py-4 text-[1.25rem] font-semibold text-white md:self-auto"
        type="submit"
      >
        button title
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-label="button title"
      class="bg-primary/90 hover:bg-primary/70 cursor-pointer w-fit self-center rounded-xl px-8 py-4 text-[1.25rem] font-semibold text-white md:self-auto"
      type="submit"
    >
      button title
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
